atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,290 @@
-                        else:
-                            logger.warning("host: %s slot: %s has invalid state", self.hostname, i)
-                            state = 'invalid'
-                            prid = 'invalid'
-                    logger.debug("%s slot %s is %s", meta['hostname'], i, state)
-                    self.slots.append({'sid': i, 'state': state, 'prid': prid})
-        except Exception as err:
-            logger.error(err)
-            logger.error("Availiability check failed fo host: %s", self.hostname)
-        return self.free_slots
-
-    def _write_free_slot(self, slot: int):
-        cmd = f'echo \'0\' > {ROOT_PATH}slot_{slot}.json'
-        data = self._ssh_run_remote_command(cmd)
-        return data
-
-    def unlock(self, slot: int, prid: str):
-        logger.debug('Unlocking')
-        retries = 0
-        while retries < 3:
-            ret = self._write_free_slot(slot)
-            if ret == b'':
-                logger.info('Unlocked')
-                return True
-            else:
-                logger.debug('Unlocking failed, retrying')
-                retries += 1
-        return False
-
-    def lock(self, slot: int, prid: str):
-        logger.debug("Aquiring lock")
-        payload = json.dumps({"prid": prid, "locked": str(datetime.now())})
-        retries = 0
-        while retries < 3:
-            filename = os.path.join(ROOT_PATH, f'slot_{slot}.json')
-            _cmd = 'if grep -qw 0 {0} ; then echo {1} > {0} ;else exit 1; fi'.format(
-                    filename,
-                    quote(payload)
-                    )
-            cmd = 'flock {} -c {}; echo $?'.format(filename, quote(_cmd))
-            logger.debug(cmd)
-            ret = self._ssh_run_remote_command(cmd)
-            if ret == b'0\n':
-                logger.info('Locked')
-                return True
-            else:
-                logger.debug('Locking failed, retrying')
-                retries += 1
-        logger.info('Locking failed')
-        raise Exception('Locking failed')
+            client.connect(
+                self.hostname, username=self.username, key_filename=self.ssh_keyfile)
+        except paramiko.ssh_exception.AuthenticationException as ex:
+            logger.warning("SSH authentication failed. %s", ex)
+        except paramiko.ssh_exception.ChannelException as ex:
+            logger.warning("Attempt to open a new Channel failed. %s", ex)
+        except paramiko.ssh_exception.NoValidConnectionsError as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        except paramiko.ssh_exception.SSHException as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        return client
 
     @backoff.on_exception(
         backoff.expo,
-        Exception,
-        factor=PARAMIKO_BACKOFF_FACTOR,
-        max_tries=PARAMIKO_MAX_RETRIES + 1,  # total tries is N retries + 1 initial attempt
+        RETRY_ON_SSH_EXCEPTIONS,
```

Comment:
Don't we have quite nested retries here?

Reply:
I hope we should not now.